packages/block-editor/src/hooks/border-color.js

```diff
@@ -0,0 +1,197 @@
+/**
+ * External dependencies
+ */
+import classnames from 'classnames';
+
+/**
+ * WordPress dependencies
+ */
+import { addFilter } from '@wordpress/hooks';
+import { getBlockSupport, hasBlockSupport } from '@wordpress/blocks';
+import { __ } from '@wordpress/i18n';
+
+/**
+ * Internal dependencies
+ */
+import ColorGradientControl from '../components/colors-gradients/control';
+import {
+	getColorClassName,
+	getColorObjectByColorValue,
+} from '../components/colors';
+import useEditorFeature from '../components/use-editor-feature';
+import { BORDER_SUPPORT_KEY } from './border';
+import { cleanEmptyObject } from './utils';
+
+const BORDER_COLOR_SUPPORT_KEY = 'color';
+const EMPTY_ARRAY = [];
+
+/**
+ * Inspector control panel containing the border color related configuration.
+ *
+ * There is deliberate overlap between the colors and borders block supports
+ * relating to border color. It can be argued the border color controls could
+ * be included within either, or both, the colors and borders panels in the
+ * inspector controls. If they share the same block attributes it should not
+ * matter.
+ *
+ * @param  {Object} props Block properties.
+ * @return {WPElement}    Border color edit element.
+ */
+export function BorderColorEdit( props ) {
+	const {
+		attributes: { borderColor, style },
+		setAttributes,
+	} = props;
+	const colors = useEditorFeature( 'color.palette' ) || EMPTY_ARRAY;
+
+	const disableCustomColors = ! useEditorFeature( 'color.custom' );
+	const disableCustomGradients = ! useEditorFeature( 'color.customGradient' );
+
+	if ( useIsBorderColorDisabled( props ) ) {
+		return null;
+	}
+
+	const onChangeColor = ( value ) => {
+		const colorObject = getColorObjectByColorValue( colors, value );
+		const newStyle = {
+			...style,
+			border: {
+				...style?.border,
+				color: colorObject?.slug ? undefined : value,
+			},
+		};
+
+		const newNamedColor = colorObject?.slug ? colorObject.slug : undefined;
+
+		setAttributes( {
+			style: cleanEmptyObject( newStyle ),
+			borderColor: newNamedColor,
+		} );
+	};
+
+	return (
+		<ColorGradientControl
+			label={ __( 'Border color' ) }
+			value={ borderColor || style?.border?.color }
+			disableCustomColors={ disableCustomColors }
+			disableCustomGradients={ disableCustomGradients }
+			onColorChange={ onChangeColor }
+		/>
+	);
+}
+
+/**
+ * Determines if there is border color support.
+ *
+ * @param  {string|Object} blockType Block name or Block Type object.
+ * @return {boolean}                 Whether there is support.
+ */
+export function hasBorderColorSupport( blockType ) {
+	const support = getBlockSupport( blockType, BORDER_SUPPORT_KEY );
+	return !! ( true === support || support?.color );
```

Comment:
This code implies that `getBlockSupport` can return either object. Is this a case?

Reply:
Yes. All border support can be opted into by simply setting the `supports.__experimentalBorder` property to true. Alternatively, individual features ( color, radius, style, and width ) can be opted into.

Reply:
I've refactored all the individual feature checks to use a single more generic function exported from `border.js`